canonical/ubuntu-pro-client
# Fix usage of PRIVATE_DIR

## Why is this needed?

This has two motivations
1. Careful review by Seth (thanks) revealed a potential race where an attacker could exchange the element we chmod to gain benefit. In this case to potentially gain execute rights. So we reduce the race by setting the permissions correctly in the mkdir() call right away. (LP: #2024204)
2. We found that there are conflicting definitions and usage of PRIVATE_DIR that cause misunderstandings and issues (like recreation of /private not having the right [too open] permissions)

## Test Steps

Usually the path is created by the postinst, which IMHO it should not do (one place is fine) see #2722 - To test the actual python code (and surprise, it was broken without a test) one needs to remove the dir created by the postinst as the python code only creates it if missing.

```
# Set up a system to be ready to be pro attached
# ensure there is no old private dir so we can check its (re-)creation
$ rm -rf /var/lib/ubuntu-advantage/private
$ strace -o verify-access.strace -rTf pro attach <redacted>
# old entries had in the log then
06:56:17.120305 mkdir("/var/lib/ubuntu-advantage/private", 0777) = 0 <0.000108>
```

The resulting file should no more have first a `mkdir` with permission `777` followed by a `chmod` to `0700`. Instead it should be created with `0700` right away.

## Checklist
 - [ ] Changes here need to be documented

## Does this PR require extra reviews?
 - [x] Yes (Sarnold)
